upload/src/addons/SV/WarningImprovements/XF/Admin/Controller/Warning.php

```diff
@@ -138,23 +158,61 @@ public function _actionAddEdit(\XF\Entity\WarningAction $action)
             $nodeRepo = $this->app()->repository('XF:Node');
             $nodes = $nodeRepo->getFullNodeList()->filterViewable();
 
-            $response->setParam('nodeTree', $nodeRepo->createNodeTree($nodes));
+            $categoryRepo = $this->getCategoryRepo();
+            $categoryTree = $categoryRepo->createCategoryTree();
+
+            $response->setParams([
+                'nodeTree' => $nodeRepo->createNodeTree($nodes),
+                'categoryTree' => $categoryTree
+            ]);
         }
 
         return $response;
     }
 
+    // underscore prefix to not be confused with actual controller actions
+    protected function _actionSaveProcess(\XF\Entity\WarningAction $action)
+    {
+        $inputFieldNames = [
+            'sv_warning_category_id' => 'uint',
+            'sv_post_node_id' => 'uint'
+        ];
+
+        $warningActionData = [];
+        foreach ($inputFieldNames AS $inputFieldName => $inputFieldFilterName)
```

Comment:
Why not directly set $action? No need from what I can see to juggle this to a generic entity preSave event and then put it in there.

Reply:
Done